### ISSUE TITLE
the terminal should not replace newlines with spaces in bash on exiting another TUI

### ISSUE THREAD

**Issue by yashpalgoyal1304** [msys2/MSYS2-packages#2741]
## Describe the issue

The terminal/shell uses spaces instead of newlines for older console history after exiting from another TUI application

## Steps to Reproduce the Problem

0. Run `bash.exe`
1. Run `vim` (specific to vim or any TUI application🤔)
2. Exit out of it
3. Resize the terminal window

- The new console history (post step 2) behaves expected
- The console history prior to step 1 now reflows i.e. behaves like it has spaces instead of new lines (BUG behaviour)
    + this become EXTREMELY hard to follow/read
    + selecting all and pasting in text editor results in same experience
- The effect can become worse when like "nested" 
like say you entered Vim, exited, resized, did some work, re-entered and exit vim, then resize again

### Use Case:
- Adding commit messages in vim while git-ting

## Additional Context: Screenshots

Pasting this prompt just for reuse, it is not relevant to the bug
```
PS1="\[\e]0;bug_record\a\]\n \[\e[33m\]\W\[\e[0m\]\n $ "
```

https://user-images.githubusercontent.com/19423063/153537371-30670492-a3c4-4dc7-917c-726b1bac8c60.mp4

## Additional Context: Operating System

- OS: Windows 10 20H2 64bit
- `mintty --version` : 3.5.1

----

<details>
<summary>
Old Report
</summary>

## Describe the issue

The terminal uses spaces instead of newlines

## Steps to Reproduce the Problem

0. Run `bash.exe`
1. Run anything that outputs things with newline characters.  \
For example, 
    + doing `cat` with files containing "enter" or say linebreak or say `/n`
    + running CMI (character mode) programmes like C programmes
    + or CLI programmes like git
    + trying downloading pip packages
    + even the command line changes line after prompt, and the command's execution result again changes line
2. Try to select post the ending of line
3. The space characters are selected
4. That partition should be empty since `/n` character should be used there
5. This can be also experienced with 
    + resizing the terminal window's width: the text reflows, which it shouldnt
    + this become EXTREMELY hard to follow/read
    + selecting all and pasting in text editor results in same experience
6. The older text in terminal is more likely to face this 5th point, the newer/recent output behaves as expected.

## Additional Context: Operating System, Screenshots

![image](https://user-images.githubusercontent.com/19423063/145941027-07425116-c7c0-46ad-9842-fc01df6ccfcb.png)

+ I can share a screenrecording video, but i dont want my username and possibly even the file location to appear there, and i dont know how to do that. so, any help would be welcome.
+ other than that, feel free to ask for any clarification, further version info etc

</details>

**Comment by yashpalgoyal1304** [comment]
coming here from https://github.com/msys2/msys2#readme

**Comment by 1480c1** [comment]
Try to pipe it into `od -c`

**Comment by yashpalgoyal1304** [comment]
* for the corresponding programme of above screenshot:
![image](https://user-images.githubusercontent.com/19423063/145949880-05c21554-1651-49da-83ca-1538660e57c7.png)
* amh, yeah, then what? what does that address? can u elaborate?
* i am guessing this was for the inspection purpose, like u wanted to test the output? right?

also just for clarification,
* i am talking about whole terminal. 
* like i am using learning stuff, 
* so, i want to copy-paste all of the interactions/terminal's text into some external file so that i can study later
* so, i cant pipe every command this way.

**Comment by 1480c1** [comment]
> * amh, yeah, then what? what does that address? can u elaborate?
> 
> * i am guessing this was for the inspection purpose, like u wanted to test the output? right?

The purpose was to confirm it only printed
```c
namaste\r\nany key press will exit this\r\n
```
which is what you actually printed from the printfs (windows converts `\n` to `\r\n` automatically in "text" mode https://docs.microsoft.com/en-us/cpp/c-runtime-library/text-and-binary-mode-file-i-o?view=msvc-170 https://docs.microsoft.com/en-us/cpp/c-runtime-library/unicode-stream-i-o-in-text-and-binary-modes?view=msvc-170)

Meaning, the blanks you were highlighting were phantom blanks, they do not exist, that is purely from the terminal, it happens in powershell as well. When you copy and paste after highlighting those lines, they won't actually get copied.

**Comment by 1480c1** [comment]
![image](https://user-images.githubusercontent.com/8345542/145952167-f85b663e-9fd9-444c-be89-4d01fa03491b.png)

![image](https://user-images.githubusercontent.com/8345542/145952197-38f4483d-b383-4a91-8a1d-776f1d2fb7a0.png)

copying and pasting this content by right clicking

```bash
C:\Users\cddeg> cat temp.c
#include <stdio.h>

int main(void) {
    printf("Hallo\n");
    printf("Next line\n");
    return 0;
}

C:\Users\cddeg> gcc temp.c
C:\Users\cddeg> ./a.exe
Hallo
Next line
C:\Users\cddeg>
```

as you may notice, there are no extra space after the `Hallo` and `Next line`

**Comment by yashpalgoyal1304** [comment]
hi, thanks for investigation

> i dont want my username and possibly even the file location to appear there, and i dont know how to do that. so, any help would be welcome.

can u share the tip for that with me, so that i can share my side

**Comment by yashpalgoyal1304** [comment]
@1480c1 I have updated the description with proper reproducible steps, title to reflect the precise issue and added a screen record as well 😊😊

**Comment by yashpalgoyal1304** [comment]
This issue is not in `mintty` but rather in windows `conhost` :(

i didnt know this, i thought that when i launch by directly typing `bash`, it opens in msys2's `mintty`

**Comment by Biswa96** [comment]
> how can i know which terminal is currently running?

Obviously, the one you are running 😄 When you click any of the mingw/clang/msys2 shortcuts it will launch mintty.

**Comment by yashpalgoyal1304** [comment]
> When you click any of the mingw/clang/msys2 shortcuts it will launch mintty.

yeah, that's what i thought while launching `bash`

![image](https://user-images.githubusercontent.com/19423063/153553740-2cc41e78-aff8-4c97-bbc6-e91ffd40f0c1.png)

**Comment by Biswa96** [comment]
I am unaware what that icon points to, it may be WSL's bash or git for Windows' bash or msys2's bash.

**Comment by yashpalgoyal1304** [comment]
> that icon points to ...

the bash, which in this case is msys2's bash. git's one is the top one, clearly distinguished.

**Comment by mmuetzel** [comment]
Are you sure it is MSYS2's `bash`? I see the following at `cmd` prompt:
```
C:\>where bash
C:\Windows\System32\bash.exe
```

**Comment by yashpalgoyal1304** [comment]
i was not sure, but i just checked:

```
$ where bash
C:\msys64\usr\bin\bash.exe
```

**Comment by yashpalgoyal1304** [comment]
```
C:\>where bash
C:\msys64\usr\bin\bash.exe
```

**Comment by mmuetzel** [comment]
There might be a reason why MSYS2's links in the start menu go through `msys2_shell.cmd` instead of calling `bash` directly. (But tbh I don't know.)
If I read that script correctly, it sets `set MSYSCON=mintty.exe` by default. That might be necessary to get a bash in mintty (instead of in conhost)...

**Comment by yashpalgoyal1304** [comment]
Just one more word of caution for future for me: octave also provides the shortcut with name `Bash Shell` which is a shortcut to `"GNU Octave\Octave-<version>\cmdshell.bat"` - aah, these ![🦇] files are haunting me now :)

[🦇]: https://github.githubassets.com/images/icons/emoji/unicode/1f987.png

**Comment by yashpalgoyal1304** [comment]
I was exploring this: https://github.com/doublecmd/doublecmd/issues/792 , and one thing crossed my mind.

> ... didnt know this, i thought that when i launch by directly typing `bash`, it opens in msys2's `mintty`
\- https://github.com/msys2/MSYS2-packages/issues/2741#issuecomment-1035942242


> There might be a reason why MSYS2's links in the start menu go through `msys2_shell.cmd`
\- https://github.com/msys2/MSYS2-packages/issues/2741#issuecomment-1036054121

* It's not that calling msys2 `bash` (the address bar show the translated address to be: `D:\msys64\usr\bin\bash.exe`) itself is calling `conhost` 
* rather, it's that being a non-gui program, windows try to open the `bash` in the default terminal - which is `conhost`
* this happens when u try to open _any_ non-gui program from explorer (say u compile simple c/++ programs and compile them, now opening the compiled program this way)
